File: crates/graphforge-bindings-py/tests/research_project.py
"""Research Project metadata and bounded local discovery (#1348)."""

from __future__ import annotations

import tempfile
import uuid

import graphforge
from graphforge.exceptions import StorageError


def _open_admitted(path: str) -> graphforge.GraphForge | None:
    try:
        return graphforge.GraphForge(path)
    except StorageError as error:
        if error.code == "GF_UNSUPPORTED_FILESYSTEM":
            return None
        raise


def check_research_metadata_and_discovery() -> None:
    with tempfile.TemporaryDirectory() as first_root, tempfile.TemporaryDirectory() as second_root:
        if _open_admitted(first_root) is None or _open_admitted(second_root) is None:
            return

        metadata = {
            "contract_version": 1,
            "title": "Arabian Nights",
            "description": None,
            "authors": [],
            "subjects": ["literature"],
            "languages": ["ar"],
            "geographic_coverage": None,
            "temporal_coverage": {
                "start": "800",
                "end": "1500",
                "label": "800-1500 CE",
            },
            "source_types": [],
            "corpus_size": None,
            "ontologies": ["narrative-events"],
            "license": None,
            "access": {
                "visibility": None,
                "access_policy": None,
                "collaborators": [],
            },
            "tags": [],
            "originating_projects": [],
            "related_projects": [],
            "canonical_identifiers": [],
            "external_identifiers": [],
            "created_at": None,
            "updated_at": None,
            "extensions": {},
            "discovery_facets": {
                "text_entry_points": 0,
                "source_entry_points": 0,
                "entity_entry_points": 0,
                "relationship_entry_points": 0,
                "story_document_entry_points": 0,
                "event_location_entry_points": 0,
                "ontology_type_entry_points": 0,
                "linguistic_property_entry_points": 0,
                "traversal_query_entry_points": 0,
                "branch_entry_points": 0,
            },
        }

        forge = graphforge.GraphForge(first_root)
        forge.update_research_metadata(
            metadata=metadata,
            operation_uuid=str(uuid.uuid4()),
        )
        reopened = graphforge.GraphForge(first_root)
        assert reopened.research_project_metadata()["title"] == "Arabian Nights"
        assert reopened.research_project_summary()["metadata"]["title"] == "Arabian Nights"

        graphforge.GraphForge(second_root).update_research_metadata(
            metadata={
                **metadata,
                "title": "Medieval Latin Corpus",
                "subjects": ["history"],
                "languages": ["la"],
                "ontologies": [],
                "temporal_coverage": None,
            },
            operation_uuid=str(uuid.uuid4()),
        )

        discovered = graphforge.GraphForge.discover_research_projects(
            project_roots=[first_root, second_root],
            languages=["ar"],
            subjects=["literature"],
            ontologies=["narrative-events"],
            temporal_label="800-1500",
        )
        assert discovered.num_rows == 1
        assert discovered.column("title").to_pylist() == ["Arabian Nights"]


def main() -> None:
    check_research_metadata_and_discovery()

File: crates/graphforge-api/tests/research_project.rs
//! Research Project metadata and bounded local discovery (#1348).

use graphforge_api::{
    DiscoverResearchProjectsRequest, GraphForge, ResearchProjectDiscoveryLimits,
    ResearchProjectDiscoveryQuery, ResearchTemporalCoverage, UpdateResearchMetadataRequest,
    WorkspaceResearchMetadata, WriteContext,
};
use graphforge_core::{GfError, ProjectErrorCode, uuid::Uuid};
use tempfile::TempDir;

fn open_admitted_graph(path: &str) -> Option<GraphForge> {
    match GraphForge::new(Some(path)) {
        Ok(graph) => Some(graph),
        Err(GfError::Project {
            code: ProjectErrorCode::UnsupportedFilesystem,
            ..
        }) => None,
        Err(error) => panic!("{error}"),
    }
}

#[test]
fn metadata_survives_reopen_and_discovery_filters_without_graph_open() {
    let first_root = TempDir::new().unwrap();
    let second_root = TempDir::new().unwrap();
    let first_path = first_root.path().to_str().unwrap();
    let second_path = second_root.path().to_str().unwrap();
    if open_admitted_graph(first_path).is_none() || open_admitted_graph(second_path).is_none() {
        return;
    }
    let mut first = GraphForge::new(Some(first_path)).unwrap();
    let _second = GraphForge::new(Some(second_path)).unwrap();

    let mut metadata = WorkspaceResearchMetadata::empty();
    metadata.title = Some("Arabian Nights".into());
    metadata.languages = vec!["ar".into()];
    metadata.subjects = vec!["literature".into()];
    metadata.ontologies = vec!["narrative-events".into()];
    metadata.temporal_coverage = Some(ResearchTemporalCoverage {
        start: Some("800".into()),
        end: Some("1500".into()),
        label: Some("800-1500 CE".into()),
    });
    first
        .update_research_metadata(UpdateResearchMetadataRequest {
            context: WriteContext {
                operation_uuid: graphforge_api::OperationId(Uuid::now_v7()),
                actor_uuid: None,
            },
            metadata,
        })
        .unwrap();

    let reopened = GraphForge::new(Some(first_path)).unwrap();
    let metadata = reopened.research_project_metadata().unwrap();
    assert_eq!(metadata.title.as_deref(), Some("Arabian Nights"));
    let summary = reopened.research_project_summary().unwrap();
    assert_eq!(summary.metadata.title.as_deref(), Some("Arabian Nights"));
    assert_eq!(summary.metadata.languages, vec!["ar".to_string()]);

    let mut second_metadata = WorkspaceResearchMetadata::empty();
    second_metadata.title = Some("Medieval Latin Corpus".into());
    second_metadata.languages = vec!["la".into()];
    second_metadata.subjects = vec!["history".into()];
    GraphForge::new(Some(second_path))
        .unwrap()
        .update_research_metadata(UpdateResearchMetadataRequest {
            context: WriteContext {
                operation_uuid: graphforge_api::OperationId(Uuid::now_v7()),
                actor_uuid: None,
            },
            metadata: second_metadata,
        })
        .unwrap();

    let request = DiscoverResearchProjectsRequest {
        project_roots: vec![
            first_root.path().to_path_buf(),
            second_root.path().to_path_buf(),
        ],
        query: ResearchProjectDiscoveryQuery {
            languages: vec!["ar".into()],
            subjects: vec!["literature".into()],
            ontologies: vec!["narrative-events".into()],
            temporal_label: Some("800-1500".into()),
            ..ResearchProjectDiscoveryQuery::default()
        },
        limits: ResearchProjectDiscoveryLimits::default(),
    };
    let discovered = GraphForge::discover_research_projects(&request).unwrap();
    assert_eq!(discovered.batches.len(), 1);
    assert_eq!(discovered.batches[0].num_rows(), 1);
    let titles = discovered.batches[0]
        .column_by_name("title")
        .unwrap()
        .as_any()
        .downcast_ref::<arrow::array::StringArray>()
        .unwrap();
    assert_eq!(titles.value(0), "Arabian Nights");
}
